spring-kafka/src/main/java/org/springframework/kafka/support/converter/ProjectingMessageConverter.java

```diff
@@ -0,0 +1,121 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.support.converter;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.lang.reflect.Type;
+import java.nio.charset.StandardCharsets;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+
+import org.springframework.core.ResolvableType;
+import org.springframework.data.projection.ProjectionFactory;
+import org.springframework.data.projection.SpelAwareProxyProjectionFactory;
+import org.springframework.data.web.JsonProjectingMethodInterceptorFactory;
+import org.springframework.kafka.support.KafkaNull;
+import org.springframework.messaging.Message;
+import org.springframework.util.Assert;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.jayway.jsonpath.spi.mapper.JacksonMappingProvider;
+
+/**
+ * A {@link MessageConverter} implementation that uses a Spring Data {@link ProjectionFactory} to bind incoming messages
+ * to projection interfaces.
+ *
+ * @author Oliver Gierke
+ * @since 2.1.1
+ */
+public class ProjectingMessageConverter extends MessagingMessageConverter {
+
+	private final ProjectionFactory projectionFactory;
+	private final MessagingMessageConverter delegate;
+
+	/**
+	 * Creates a new {@link ProjectingMessageConverter} using the given {@link ObjectMapper}.
+	 *
+	 * @param mapper must not be {@literal null}.
+	 */
+	public ProjectingMessageConverter(ObjectMapper mapper) {
+
+		Assert.notNull(mapper, "ObjectMapper must not be null!");
+
+		JacksonMappingProvider provider = new JacksonMappingProvider(mapper);
+		JsonProjectingMethodInterceptorFactory interceptorFactory = new JsonProjectingMethodInterceptorFactory(provider);
+
+		SpelAwareProxyProjectionFactory factory = new SpelAwareProxyProjectionFactory();
+		factory.registerMethodInvokerFactory(interceptorFactory);
+
+		this.projectionFactory = factory;
+		this.delegate = new StringJsonMessageConverter(mapper);
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see org.springframework.kafka.support.converter.MessagingMessageConverter#convertPayload(org.springframework.messaging.Message)
+	 */
+	@Override
+	protected Object convertPayload(Message<?> message) {
+		return this.delegate.convertPayload(message);
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see org.springframework.kafka.support.converter.MessagingMessageConverter#extractAndConvertValue(org.apache.kafka.clients.consumer.ConsumerRecord, java.lang.reflect.Type)
+	 */
+	@Override
+	protected Object extractAndConvertValue(ConsumerRecord<?, ?> record, Type type) {
+
+		Object value = record.value();
+
+		if (value == null) {
+			return KafkaNull.INSTANCE;
+		}
+
+		Class<?> rawType = ResolvableType.forType(type).resolve(Object.class);
+
+		if (!rawType.isInterface()) {
+			return this.delegate.extractAndConvertValue(record, type);
+		}
+
+		InputStream inputStream = new ByteArrayInputStream(getAsByteArray(value));
```

Comment:
While you here, @olivergierke , I would like to ask on this line:

Is it OK to keep `InputStream` opened or would be better to do this:
```
try (InputStream inputStream = new ByteArrayInputStream(getAsByteArray(value))) {
	return this.projectionFactory.createProjection(rawType, inputStream);
}
catch (IOException e) {
	throw new ConversionException("Cannot project from the value", e);
}
```
?

Test are passed and I see that final object for method invocation is the `Map` based on the JSON parsing.
So, sounds like Jackson has finished its parsing process already and we have to close `InputStream`

Your thoughts?

Thanks.

Reply:
Well, actually we don't need to do that. The `JacksonJsonNodeJsonProvider` performs this:
```
public JsonNode readTree(Reader r) throws IOException {
       return _readTreeAndClose(_jsonFactory.createParser(r));
}
```

So, nevermind 😄 
Merging...

Reply:
A `ByteArrayInputStream` doesn't need to be closed explicitly as its `close()` method is a no-op anyway.